subsys/net/lib/dns/dns_sd.h

```diff
@@ -0,0 +1,141 @@
+ * @param service buffer to store the null-terminated service
+ * @param service_size the size of @p service
+ * @param proto buffer to store the null-terminated proto
+ * @param proto_size the size of @p proto
+ * @param domain buffer to store the null-terminated domain
+ * @param domain_size the size of @p domain
+ * @return on success, a positive number representing length of the query
+ * @return on failure, a negative errno value
+ */
+int dns_sd_extract_service_proto_domain(const uint8_t *query,
+	size_t query_size, struct dns_sd_rec *record, char *service,
+	size_t service_size, char *proto, size_t proto_size,
+	char *domain, size_t domain_size);
+
+/**
+ * @brief See if the DNS SD @p filter matches the @p record
+ *
+ * The fields in @p filter should be populated with filter elements to
+ * identify a possible match. If pointer fields are set to NULL, they
+ * act as a wildcard in the matching process. I.e. they will match
+ * anything. Similarly, the @ref dns_sd_rec.port field may be set to 0
+ * to be used as a wildcard.
+ *
+ * The @ref dns_sd_rec.text and @ref dns_ds_rec.text_size fields
+ * are not included in the matching process.
+ *
+ * For example, the filter below can be used to match any
+ * "_http._tcp" records.
+ *
+ * @code{c}
+ * const struct dns_sd_rec *it;
+ * struct dns_sd_rec filter = {
+ *     // match any instance
+ *     .instance = NULL,
+ *     // match records with service "_http"
+ *     .service = "_http",
+ *     // match records with protocol "_tcp"
+ *     .proto = "_tcp",
+ *     // match any domain
+ *     .domain = NULL,
+ *     // match any port
+ *     .port = 0,
+ * };
+ *
+ * DNS_SD_FOREACH(it) {
+ *     if (dns_sd_rec_match(it, filter)) {
+ *         // found a match!
+ *     }
+ * }
+ * @endcode{c}
+ *
+ * @param record The reference DNS-SD record
+ * @param filter The DNS-SD record filter
+ *
+ * @return true if the @p record matches the @p filter
+ * @return false if @p record is not a match for @p filter
+ * @return false if either @p record or @p filter are invalid
+ */
+bool dns_sd_rec_match(const struct dns_sd_rec *record,
+	const struct dns_sd_rec *filter);
+
+/**
+ * @brief Handle a DNS PTR Query with DNS Service Discovery
+ *
+ * This function should be called once for each DNS-SD record that
+ * matches a particular DNS PTR query.
+ *
+ * If there is no IPv4 address to advertise, then @p addr4 should be
+ * NULL.
+ *
+ * If there is no IPv6 address to advertise, then @p addr6 should be
+ * NULL.
+ *
+ * @param inst the DNS-SD record for to advertise
+ * @param addr4 pointer to the IPv4 address
+ * @param addr6 pointer to the IPv6 address
+ * @param buf output buffer
+ * @param buf_size size of the output buffer
+ *
+ * @return on success, number of bytes written to @p buf
+ * @return on failure, a negative errno value
+ */
+int dns_sd_handle_ptr_query(const struct dns_sd_rec *inst,
+	const struct in_addr *addr4, const struct in6_addr *addr6,
```

Comment:
Instead of these two variables, there is the `struct sockaddr` that can hold both address types. We use sockaddr in other places when IP address can be either type.

Reply:
The spec says that both IPv4 and IPv6 address records should be sent, if available, so we allow both. If one of the addresses is NULL, then the implication is that it is not available.

Reply:
> The spec says that both IPv4 and IPv6 address records should be sent, if available, so we allow both.

Ok, makes sense.